tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -3411,6 +3411,13 @@ SpirvEmitter::processACSBufferAppendConsume(const CXXMemberCallExpr *expr) {
     // Note that we are returning a pointer (lvalue) here inorder to further
     // acess the fields in this element, e.g., buffer.Consume().a.b. So we
     // cannot forcefully set all normal function calls as returning rvalue.
+    if (isBoolOrVecOfBoolType(bufferElemTy)) {
```

Comment:
based on the above comment, the `Consume()` method should actually be returning a pointer (lvalue), so that the caller can use it for further indexing if necessary. But after your change, the `castToBool` result is an rvalue.

for example, the following test case does not work with your code:
```hlsl
ConsumeStructuredBuffer<bool2> foo;
AppendStructuredBuffer<bool> bar;

void main() {
  bar.Append(foo.Consume().x);
}
```

I think the fix might be by updating the `if(isAppend)` clause above?

Note that the second test (`type.consume-structured-buffer.bool.hlsl`) passes even without your change. I'm not saying remove the test (the more tests the better). But just pointing out.

You may also want to add a test that exercises boolean in a struct. e.g.
```hlsl
struct S {
  float4 a;
  bool   b;
};

ConsumeStructuredBuffer<S> foo;
AppendStructuredBuffer<bool> bar;
ConsumeStructuredBuffer<S> zoo;

void main() {
  bar.Append(foo.Consume().b);
  zoo.Append(foo.Consume());
}
```

Reply:
Updated. PTAL.